enhancements/operator-conditions.md

```diff
@@ -0,0 +1,180 @@
+
+>An operator installed with automatic updates is currently performing a migration that could result in data loss if OLM upgrades the operator. While this migration process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the migration process to finish, resulting in unrecoverable data loss. Had OLM given the operator the ability to communicate that it should not be upgraded, no data loss would have occurred.
+
+As shown in the example above, by allowing operators to report complex condition and influence OLM's management, the overall user experience could be improved for both operator authors and OLM users.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate conditions that influences OLM behavior.
+- Allow the list of "OLM Supported Conditions" to grow in the future.
+- Update OLM to support the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior if the operator does not take advantage of this feature.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want operators managed by OLM to be able to communicate complex states that OLM cannot infer on its own.
+
+#### Story 2
+
+As an operator author, I want operators managed by OLM to be able to communicate whether or not they are ready to be upgraded.
+
+## Design Details
+
+### OperatorHealth CRD
+
+OLM will create a channel for operators to communicate complex conditions by introducing the `OperatorHealth` [CustomResourceDefinition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+When an operator is installed, OLM will create an `OperatorHealth` CustomResource (CR) for the operator. Operators will use a library provided by the [operator-sdk](https://github.com/operator-framework/operator-sdk) team to set [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions) on the CR to communicate complex states to OLM. For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:
```

Comment:
Will OLM prevent operators from requesting OperatorHealth create permission?

Reply:
That's a good point - I will add a bullet under Risk and Mitigations that calls out that OLM will fails CSV installs that request CREATE or DELETE RBAC for OperatorHealth CRs.

Reply:
Added here: https://github.com/operator-framework/enhancements/pull/42/files#diff-2258f1c6f47041cff3143c16e4a336afR181